src/services/EdFi.DataManagementService.Api/Core/Middleware/ApiSchemaValidationMiddleware.cs

```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using EdFi.DataManagementService.Api.Core.ApiSchema;
+using EdFi.DataManagementService.Core.Pipeline;
+
+namespace EdFi.DataManagementService.Api.Core.Middleware;
+
+public class ApiSchemaValidationMiddleware(
+    IApiSchemaProvider _apiSchemaProvider,
+    IApiSchemaValidator _apiSchemaValidator,
+    ILogger _logger
+) : IPipelineStep
+{
+    public async Task Execute(PipelineContext context, Func<Task> next)
+    {
+        _logger.LogDebug(
+            "Entering ApiSchemaValidationMiddleware- {TraceId}",
+            context.FrontendRequest.TraceId
+        );
+
+        var document = _apiSchemaProvider.ApiSchemaRootNode;
+        var validationErrors = _apiSchemaValidator.Validate(document).Value;
```

Comment:
Instead of the AspNetCore stuff, do a simple cache of the `validationErrors` result of the `Validator()` call in an instance variable initialized to `null`. If null, call `Validator()`. If not, do what you are doing.

I think that's all you need to self-contain this in Core.

Reply:
Actually, you'll want it to be an instance variable Lazy. I always forget about how complicated threading is in C#.